src/main/java/org/candlepin/subscriptions/marketplace/MarketplaceSubscriptionIdProvider.java

```diff
@@ -0,0 +1,120 @@
+        SubscriptionRepository subscriptionRepo, SubscriptionSyncController syncController,
+        ProductProfileRegistry profileRegistry) {
+        this.collector = collector;
+        this.subscriptionRepo = subscriptionRepo;
+        this.syncController = syncController;
+        this.profileRegistry = profileRegistry;
+    }
+
+    public Optional<Object> findSubscriptionId(String accountNumber, Key usageKey,
+        OffsetDateTime rangeStart, OffsetDateTime rangeEnd) {
+        Assert.isTrue(Usage._ANY != usageKey.getUsage(), "Usage cannot be _ANY");
+        Assert.isTrue(ServiceLevel._ANY != usageKey.getSla(), "Service Level cannot be _ANY");
+
+        String productId = usageKey.getProductId();
+        ProductProfile profile = profileRegistry.findProfileForSwatchProductId(productId);
+        Set<String> roles = profile.getRolesBySwatchProduct().getOrDefault(productId, Collections.emptySet());
+
+        List<Subscription> result = fetchSubscriptions(accountNumber, usageKey, roles, rangeStart, rangeEnd);
+
+        if (result.isEmpty()) {
+            /* If we are missing the subscription, call out to the MarketplaceSubscriptionCollector
+               to fetch from Marketplace.  Sync all those subscriptions. Query again. */
+            var subscriptions = collector.fetchSubscription(accountNumber, usageKey);
+            subscriptions.forEach(syncController::syncSubscription);
+            result = fetchSubscriptions(accountNumber, usageKey, roles, rangeStart, rangeEnd);
+        }
+
+        if (result.isEmpty()) {
+            return Optional.empty();
+        }
+
+        if (result.size() > 1) {
+            log.warn("Multiple subscriptions found for account {} with key {} and roles {}. Selecting first" +
+                " result", accountNumber, usageKey, roles);
+        }
+        return Optional.of(result.get(0).getMarketplaceSubscriptionId());
+    }
+
+    protected List<Subscription> fetchSubscriptions(String accountNumber, Key usageKey, Set<String> roles,
+        OffsetDateTime rangeStart, OffsetDateTime rangeEnd) {
+        List<Subscription> result =
+            subscriptionRepo.findSubscriptionByAccountAndUsageKey(accountNumber, usageKey, roles);
+        result = filterByDateRange(result, rangeStart, rangeEnd);
+        result = filterByMissingMarketplaceSubscriptionId(result);
+        return result;
+    }
+
+    private List<Subscription> filterByMissingMarketplaceSubscriptionId(List<Subscription> result) {
+        return result.stream()
+            .filter(x -> StringUtils.hasText(x.getMarketplaceSubscriptionId()))
+            .collect(Collectors.toList());
+    }
+
+    private List<Subscription> filterByDateRange(List<Subscription> result, OffsetDateTime rangeStart,
+        OffsetDateTime rangeEnd) {
+        return result.stream()
+            .filter(x -> rangeStart.isBefore(x.getStartDate()) && rangeEnd.isAfter(x.getEndDate()))
```

Comment:
```suggestion
        return result.stream().filter(
            x -> (rangeStart.isEqual(x.getStartDate()) || rangeStart.isAfter(x.getStartDate())) &&
            (rangeEnd.isEqual(x.getEndDate()) || rangeEnd.isBefore(x.getEndDate())))
```

We need to handle the edges of the ranges appropriately... Crude ascii:
```
1. snapshot range neatly inside:
| subscription      |
  | snapshot range |

rangeStart.isAfter(x.startDate) == true
rangeEnd.isBefore(x.endDate) == true

2. snapshot range "just fits"...
| subscription       |
| snapshot range     |

x.startDate == rangeStart
x.endDate == rangeEnd
```
(Other permutations left as an exercise for the reader).